### PR TITLE
RemoteCreate/Update override

### PR DESCRIPTION
**Pull request by nireeshT**
RemoteCreate/Update override

### CODE REVIEW

File: src/main/java/com/capitalone/dashboard/service/DashboardRemoteServiceImpl.java [hunk 1]
```diff
@@ -134,26 +138,24 @@ public Dashboard remoteCreate(DashboardRemoteRequest request, boolean isUpdate)
...
-                list.addAll( component.getCollectorItems(CollectorType.StaticSecurityScan ) );
-                list.addAll( component.getCollectorItems(CollectorType.CodeQuality ) );
-                list.addAll( component.getCollectorItems(CollectorType.LibraryPolicy ) );
-
-                List< ObjectId > collectorItemIdList = list.stream().map( CollectorItem::getId).collect(Collectors.toList() );
-                widgetRequest.getCollectorItemIds().addAll( collectorItemIdList );
-            }
+            widgetRequest.getCollectorItemIds().forEach(id->{
+                CollectorItem  c = collectorItemRepository.findOne(id);
+                Collector collector = collectorRepository.findOne(c.getCollectorId());
```
Comment (rvema) [inline src/main/java/com/capitalone/dashboard/service/DashboardRemoteServiceImpl.java]: How do we handle bad data? assume if the collector item does not exist, then whole process will be aborted. You want to reconsider the logic

File: src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java [hunk 2]
```diff
@@ -273,16 +283,27 @@ private boolean isSafeTeamDashboardDelete(Dashboard dashboard) {
...
+        return component;
+    }
+
+    @Override
+    public Component associateCollectorToComponent(ObjectId componentId, List<ObjectId> collectorItemIds,Component component) {
+        if (componentId == null || collectorItemIds == null) {
+            // Not all widgets gather data from collectors
+            return null;
+        }
+        associateCollectorItemsToComponent(collectorItemIds, false, component);
```
Comment (rvema) [inline src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java]: need to check the logic?

File: src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java [hunk 3]
```diff
@@ -645,24 +657,62 @@ public void deleteWidget(Dashboard dashboard, Widget widget,ObjectId componentId
...
+        if(widgetName.equalsIgnoreCase(PERFORMANCE)) return CollectorType.AppPerformance;
+        if(widgetName.equalsIgnoreCase(CLOUD)) return CollectorType.Cloud;
+        if(widgetName.equalsIgnoreCase(CHATOPS)) return CollectorType.ChatOps;
+        if(widgetName.equalsIgnoreCase(TEST)) return CollectorType.Test;
+        return null;
+    }
+
+    private static String findWidgetName(CollectorType collectorType){
+        if(collectorType.equals(CollectorType.Build)) return BUILD;
+        if(collectorType.equals(CollectorType.AgileTool)) return FEATURE;
```
[the remark this answers is not in the document]
Comment (nireeshT) [reply src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java]: @rvema  I thought early exits are much better and less code.:) . I agree switch is more maintainable.

File: src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java [hunk 4]
```diff
@@ -645,24 +657,62 @@ public void deleteWidget(Dashboard dashboard, Widget widget,ObjectId componentId
 
     }
 
+    private void dashboardUpdate(Dashboard dashboard, int index) {
+        if(index!=-1){
+            dashboard.getWidgets().set(index, null);
+            List<Widget> widgets = dashboard.getWidgets();
+            List<Widget> updatedWidgets = new ArrayList<>();
+            for (Widget w: widgets) {
```
[the remark this answers is not in the document]
Comment (nireeshT) [reply src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java]: done

File: src/main/java/com/capitalone/dashboard/service/DashboardRemoteServiceImpl.java (hunk 1 above)
Comment (nireeshT) [reply src/main/java/com/capitalone/dashboard/service/DashboardRemoteServiceImpl.java]: added null-safe

**Pull request merged**